wire-ios-data-model/Source/Model/Message/ConversationMessage+Reaction.swift

```diff
@@ -18,111 +18,113 @@
+            emojis: reactions,
+            messageID: messageID
+        ))
 
-        let emoji = unicodeValue ?? ""
-        let reaction = WireProtos.Reaction.createReaction(emoji: emoji, messageID: messageID)
-        let genericMessage = GenericMessage(content: reaction)
+        do {
+            let clientMessage = try conversation.appendClientMessage(
+                with: genericMessage,
+                expires: false,
+                hidden: true
+            )
 
-        guard let conversation = message.conversation else { return nil }
+            message.setReactions(
+                reactions,
+                forUser: .selfUser(in: context)
+            )
 
-        do {
-            let clientMessage = try conversation.appendClientMessage(with: genericMessage, expires: false, hidden: true)
-            message.addReaction(unicodeValue, forUser: .selfUser(in: context))
             return clientMessage
         } catch {
             Logging.messageProcessing.warn("Failed to append reaction. Reason: \(error.localizedDescription)")
             return nil
         }
     }
 
-    @discardableResult
-    @objc public static func addReaction(_ reaction: MessageReaction, toMessage message: ZMConversationMessage) -> ZMClientMessage? {
-        // confirmation that we understand the emoji
-        // the UI should never send an emoji we dont handle
-        if Reaction.transportReaction(from: reaction.unicodeValue) == .none {
-            fatal("We can't append this reaction \(reaction.unicodeValue), this is a programmer error.")
+    @objc public func setReactions(
+        _ reactions: Set<String>,
+        forUser user: ZMUser
+    ) {
+        // Remove all existing reactions for this user.
+        for reaction in self.reactions where reaction.users.contains(user) {
+            reaction.mutableSetValue(forKey: ZMReactionUsersValueKey).remove(user)
         }
 
-        return appendReaction(reaction.unicodeValue, toMessage: message)
-    }
-
-    @objc public static func removeReaction(onMessage message: ZMConversationMessage) {
-        _ = appendReaction(nil, toMessage: message)
-    }
-
-    @objc public func addReaction(_ unicodeValue: String?, forUser user: ZMUser) {
-        removeReaction(forUser: user)
-        guard let unicodeValue = unicodeValue,
-            unicodeValue.count > 0 else {
+        // Add all new reactions for this user.
+        for reaction in reactions {
+            if let existingReaction = self.reactions.first(where: {
+                $0.unicodeValue == reaction
+            }) {
+                existingReaction.mutableSetValue(forKey: ZMReactionUsersValueKey).add(user)
+            } else if Reaction.validate(unicode: reaction) {
+                let newReaction = Reaction.insertReaction(
+                    reaction,
+                    users: [user],
+                    inMessage: self
+                )
+
+                self.mutableSetValue(forKey: "reactions").add(newReaction)
```

Comment:
This may be easier to read? Up to you, I don't have a strong opinion on this one 

```suggestion
        // Add all new reactions for this user.
        for reaction in reactions {
            let existingReaction = self.reactions.first(where: {
                $0.unicodeValue == reaction
            })
            
            if let existingReaction = existingReaction {
                existingReaction.mutableSetValue(forKey: ZMReactionUsersValueKey).add(user)
            } else if Reaction.validate(unicode: reaction) {
                let newReaction = Reaction.insertReaction(
                    reaction,
                    users: [user],
                    inMessage: self
                )

                mutableSetValue(forKey: "reactions").add(newReaction)
```